### PR TITLE
[OpenCL] Fix profiling hang for OpenCL device

### PR DESCRIPTION
**Pull request by valmat07**
Profiling for OpenCL runs was hanging because nested timer runs were not correctly handled. This is because each call to the `Timer::Start` function on OpenCL device causes the OpenCL event queue to be cleared.

The hang was due to the fact that the minimum run time was not reached for profiling, if it is set greater than zero. Since the queue is cleared every `Start` call, the time of only one run is returned, not `number`.

It happens in this cycle:

https://github.com/apache/tvm/blob/75ec1cffa9f160dd3165fedbf4408731ebfa797a/src/runtime/profiling.cc#L876-L891

Here `pf.CallPacked(args, &temp)` call function which contains one more call `Timer::Start` and `Timer::End`.

Also profiling was hanging for `__nop` functions like `reshape` due to there are no OpenCL calls for this node. So that OpenCL timer returns 0, which is less than the minimum time value.

The `__nop` functions are now skipped for profiling.

### CODE REVIEW

**Review by echuraev — CHANGES_REQUESTED** [review]
Will this code work in the following case?
Pseudocode:
```
Timer.ctor()
for (i: 0 -> 10) {
    Timer.Start();
    event1
    event2
    event3
    Timer.Stop();
}
```

After the first iteration and `Stop` call, the `event_start_idxs[0]` will be equal to `3`. On the second iteration `count_timer_execs == 0`. We will recreate a command queue but `event_start_idxs.size()` will be equal to 1. So, we will not go to the condition `event_start_idxs.size() < count_timer_execs` and the `event_start_idxs[0]` won't be reinitialized. Am I right?

Also, is it real scenario, or probably I have imagined something unreal?

**Comment by echuraev** [comment]
@valmat07 Thank you for your reply! Yes, I see, it looks like it should work. In this case, could you please add a test which will illustrate the problem which you have fixed in this PR? You can add test here: https://github.com/apache/tvm/tree/main/tests/cpp-runtime/opencl
I think that test can be simple as this one:
https://github.com/apache/tvm/blob/9963b59ffa489db61358dedd35a2453a5ca666b9/tests/cpp/profiling_test.cc#L28-L39

File: tests/cpp-runtime/opencl/opencl_timer_test.cc [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    OPENCL_CHECK_ERROR(err);
+    OPENCL_CALL(clEnqueueWriteBuffer(queue, cl_buf, false, 0, BUFF_SIZE * sizeof(cl_int), tmp_buf, 0, NULL, &ev));
+    OPENCL_CALL(clReleaseMemObject(cl_buf));
+    workspace->events[thr->device.device_id].push_back(ev);
+    nested_timer->Stop();
+    nested_time_sum += nested_timer->SyncAndGetElapsedNanos();
+  }
+  init_timer->Stop();
+
+  free(tmp_buf);
```
Comment (echuraev) [inline tests/cpp-runtime/opencl/opencl_timer_test.cc:58]: You have allocated this object with `new`, so you must use `delete` instead of `free`.

**Review by echuraev — CHANGES_REQUESTED** [review]

**Review by echuraev — APPROVED** [review]
LGTM. Thanks

**Comment by echuraev** [comment]
@masahi could you please take a look?

**Review by masahi — APPROVED** [review]

**Pull request merged**